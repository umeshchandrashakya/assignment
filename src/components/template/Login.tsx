import React from "react";
import styled from "styled-components";
import InputText from "../atoms/InputText";
import Label from "../atoms/Label";
import Field from "../molecules/Field";
import Form from "../organisms/LoginForm";
import img3 from "../../assets/img3.jpg";
import { StyledSpinner } from "../atoms/StyledSpinner";

const StyledLogin = styled.div`
  display: flex;
  align-items: center;
  flex-flow: column;
  /* width: 350px;
  height: 350px; */
  margin: 0 auto;
  background-repeat: no-repeat;
  background-image: url(${img3});
  height: 100vh;
  width: 100vw;
  background-size: cover;
  background-position: center center;
  box-sizing: border-box;
  min-height: 30em;
  min-width: max-content;
  position: relative;
  /* min-width: 25em; */
  h2 {
    font-family: Arial, Helvetica, sans-serif;
    font-size: 14px;
  }
  .spinner-container {
    width: 100vw;
    height: 100vh;
    z-index: 1;
    position: absolute;
    justify-content: center;
    display: flex;
    align-items: center;
  }
`;

interface Props {
  title: string;
  isLoading: boolean;
  onUsernameChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onPasswordChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onClick: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  userNameEmpty: boolean;
  passwordEmpty: boolean;
  onSignupClick: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => void;
}

const LoginForm = (props: Props) => (
  <StyledLogin className="container-fluid">
    <div
      className="spinner-container"
      style={{ visibility: props.isLoading ? "visible" : "hidden" }}
    >
      <StyledSpinner></StyledSpinner>
    </div>
    <Form {...props} />
  </StyledLogin>
);

export default LoginForm;
