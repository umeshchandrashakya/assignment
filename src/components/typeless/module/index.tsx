import React from "react";
import { LoginActions, LoginState, loginHandle } from "../interface/login";
import { RegActions, RegState, regHandle } from "../interface/reg";
import api from "../../../apis";
import history from "../../../history";
import App from "../../App";

// Create Epic for side effects
loginHandle
  .epic()
  // Listen for `count` and dispatch `countDone` with 500ms delay
  .on(LoginActions.signIn, ({ email, password }) => {
    //Rx.of(CounterActions.countDone(1)).pipe(Rx.delay(1000))
    return signIn(email, password);
  });

const initialStateLogin: LoginState = {
  isLoading: false,
  isSignedIn: false,
  result: { message: "", data: [] },
  error: {},
  username: "",
  password: "",
  formSubmitStatus: "",
  activeItem: "home",
  userNameEmpty: false,
  passwordEmpty: false
};

// Create a reducer
// It's compatible with a standard reducer `(state, action) => state`
// Under the hood it uses `immer` and state mutations are allowed
loginHandle
  .reducer(initialStateLogin)
  .on(LoginActions.signIn, state => {
    state.isLoading = true;
  })
  .on(LoginActions.signInSuccess, (state, { result }) => {
    state.isLoading = false;
    state.isSignedIn = true;
    state.result = result;
    state.formSubmitStatus = "success";
    history.push("/");
    state.username = "";
    state.password = "";
    console.log("Login Suceess");
  })
  .on(LoginActions.signInError, (state, { error }) => {
    state.isLoading = false;
    state.formSubmitStatus = "error";
    state.error = error;
    console.log("Login Error");
  })
  .on(LoginActions.usernameChange, (state, { username }) => {
    state.username = username;
  })
  .on(LoginActions.passwordChange, (state, { password }) => {
    state.password = password;
  })
  .on(LoginActions.activeItemChange, (state, { activeItem }) => {
    state.activeItem = activeItem;
  })
  .on(LoginActions.isUserNameEmpty, (state, { isField }) => {
    state.userNameEmpty = isField;
  })
  .on(LoginActions.isPasswordEmpty, (state, { isField }) => {
    state.passwordEmpty = isField;
  });

export const signIn = async (email: string, password: string) => {
  try {
    const response = await api.post("/api/user/login", {
      email: email,
      password: password
    });

    return LoginActions.signInSuccess(response.data);
  } catch (error) {
    return LoginActions.signInError(error);
  }
};

// Create Epic for side effects
regHandle
  .epic()
  // Listen for `count` and dispatch `countDone` with 500ms delay
  .on(RegActions.register, ({ name, email, phone, password }) => {
    //Rx.of(CounterActions.countDone(1)).pipe(Rx.delay(1000))
    return register(name, email, phone, password);
  });

const initialStateReg: RegState = {
  isLoading: false,
  isSuccess: false,
  result: { message: "", data: [] },
  error: {},
  name: "",
  email: "",
  phone: "",
  password: "",
  formSubmitStatus: "",
  activeItem: "home"
};

// Create a reducer
// It's compatible with a standard reducer `(state, action) => state`
// Under the hood it uses `immer` and state mutations are allowed
regHandle
  .reducer(initialStateReg)
  .on(RegActions.register, state => {
    state.isLoading = true;
  })
  .on(RegActions.regSuccess, (state, { result }) => {
    state.isLoading = false;
    state.isSuccess = true;
    state.result = result;
    state.formSubmitStatus = "success";
    history.push("/");
  })
  .on(RegActions.regError, (state, { error }) => {
    state.isLoading = false;
    state.formSubmitStatus = "error";
    state.error = error;
  })
  .on(RegActions.nameChange, (state, { name }) => {
    state.name = name;
  })
  .on(RegActions.emailChange, (state, { email }) => {
    state.email = email;
  })
  .on(RegActions.phoneChange, (state, { phone }) => {
    state.phone = phone;
  })
  .on(RegActions.passwordChange, (state, { password }) => {
    state.password = password;
  })
  .on(RegActions.activeItemChange, (state, { activeItem }) => {
    state.activeItem = activeItem;
  });

export const register = async (
  name: string,
  email: string,
  phone: string,
  password: string
) => {
  try {
    const response = await api.post("/api/user", {
      name: name,
      email: email,
      phone: phone,
      password: password
    });
    return RegActions.regSuccess(response.data);
  } catch (error) {
    return RegActions.regError(error);
  }
};

// Entry point component for this module
const BaseModule: React.FC = () => {
  // load epic and reducer
  loginHandle();
  regHandle();

  return <App />;
};

export default BaseModule;
