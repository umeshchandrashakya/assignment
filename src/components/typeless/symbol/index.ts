export const LoginSymbol = Symbol('login');
export const RegistrationSymbol = Symbol('registration');