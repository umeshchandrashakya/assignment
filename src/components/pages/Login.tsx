import React from "react";
import { useActions } from "typeless";
import { LoginActions, getLoginState } from "../typeless/interface/login";

const Login: React.FC = () => {
  // wrap actions with `dispatch`
  const { usernameChange, passwordChange, signIn } = useActions(LoginActions);
  // get state from store
  const { username, password, isLoading, error, isSignedIn, formSubmitStatus } = getLoginState.useState();

  const onLoginClicked = () => {
    signIn(username, password);
  }

  const getStatus = () => {
     if (formSubmitStatus) {
       return isLoading ? 'Loading, please wait...' : (isSignedIn ? 'Successfully logged in' : `Error: ${error}`)
     }
     return '';
  }

  return (
    <div>
      <h1>Login</h1>
      <form>
        <input className='input1' value={username} placeholder='Username' onChange={e => usernameChange(e.target.value)} />
        <input className='input2' value={password} placeholder='Password' onChange={e => passwordChange(e.target.value)} />
        <button type='button' onClick={onLoginClicked}>Login</button>
      </form>
      <h3>{getStatus()}</h3>
    </div>
  );
};

export default Login;
