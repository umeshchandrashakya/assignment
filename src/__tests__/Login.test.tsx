import React from 'react';
import Login from '../components/pages/Login';
import * as enzyme from 'enzyme';
//jest.unmock('axios');
import axios from 'axios';
import { DefaultTypelessProvider } from 'typeless';
//import MockAdapter from 'axios-mock-adapter';
//import { userLogin, getUsers } from '../apis';

// wrap the module in jest.mock() 
jest.mock('axios');
// This sets the mock adapter on the default instance
//var mock = new MockAdapter(axios);
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('<Login />', () => {

    let wrapper: enzyme.ShallowWrapper;

    beforeEach(() => {
        wrapper = enzyme.shallow(
            <DefaultTypelessProvider>
                <Login />
            </DefaultTypelessProvider>
        );
    });

    it('should call a userLogin function', async () => {
        const data = {
            email: 'brij@hcl.com',
            password: '123456'
        }
        const responseData = { data: { token: '' } };
        //mock.onPost('https://reqres.in/api/login').reply(200,responseData);
        mockedAxios.get.mockImplementationOnce(() => Promise.resolve(responseData))
        //mockedAxios.get.mockImplementationOnce(()={})
        //mockedAxios.post.mockResolvedValue(responseData);
        //const consoleLogSpy = jest.spyOn(console, 'log');
        const stateCallback = jest.fn();
        const response = await axios.get('https://simple-login-signup-rest.herokuapp.com/api/user');
        //const response = await getUsers();
        expect(response).toEqual(responseData);
        expect(mockedAxios.get).toHaveBeenCalledTimes(1);
        expect(mockedAxios.get).toHaveBeenCalledWith('https://simple-login-signup-rest.herokuapp.com/api/user');
        //expect(wrapper.find('success')).not.toBeNull();
        //done();
    });

    it('should show login success', () => {
        //const wrapper = enzyme.shallow(<Login />);

        //expect(wrapper.find('Test')).toMatchSnapshot();

        //jest.spyOn(user, '')

        const usernameInput = wrapper.find('.input1');
        usernameInput.simulate('change', { target: { value: 'brij@hcl.com' } });

        const passwordInput = wrapper.find('.input2');
        passwordInput.simulate('change', { target: { value: '123456' } });

        const data = {
            email: 'brij@hcl.com',
            password: '123456'
        }
        const responseData = { data: { message: 'success', data: {} } };
        //mock.onPost('https://reqres.in/api/login').reply(200,responseData);
        mockedAxios.post.mockImplementationOnce(() => Promise.resolve(responseData))

        const button = wrapper.find('button');
        button.simulate('click');

        expect(mockedAxios.post).toHaveBeenCalledTimes(1);
        expect(mockedAxios.post).toHaveBeenCalledWith('https://simple-login-signup-rest.herokuapp.com/api/user/login', data);

    })

    afterEach(() => {
        jest.clearAllMocks();
    });
})

