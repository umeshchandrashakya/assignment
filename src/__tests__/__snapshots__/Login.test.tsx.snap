// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render Login component correctly 1`] = `ShallowWrapper {}`;
